### PR TITLE
Support PushTopic replay extension

### PR DESCRIPTION
**Pull request by sirlatrom**
See Salesforce API documentation here: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/replay_pushtopic_events.htm

Signed-off-by: Sune Keller <absukl@almbrand.dk>

### CODE REVIEW

**Comment by sirlatrom** [comment]
Ping @davsclaus

**Comment by davsclaus** [comment]
There is merge conflicts now, can you rebase this PR ?

**Comment by sirlatrom** [comment]
@davsclaus Rebased, also fixed a number of checkstyle errors.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/api/dto/SObject.java [hunk 1]
```diff
@@ -43,6 +43,7 @@
     private Boolean undeletable;
     private Boolean triggerable;
     private Boolean compactLayoutable;
+    private Boolean mruEnabled;
```
Comment (sirlatrom) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/api/dto/SObject.java]: This one came back as a deserialization error when setting the Salesforce API version to 37.0.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/api/dto/SObjectUrls.java [hunk 2]
```diff
@@ -31,6 +31,7 @@
     private String compactLayouts;
     private String caseRowArticleSuggestions;
     private String push;
+    private String defaultValues;
```
Comment (sirlatrom) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/api/dto/SObjectUrls.java]: This one came back as a deserialization error when setting the Salesforce API version to 37.0.

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java [hunk 3]
```diff
@@ -143,6 +147,12 @@
     @UriParam
     private String instanceId;
 
+    // Streaming API properties
+    @UriParam
+    private Integer defaultReplayId;
+    @UriParam
+    private Map<String, Integer> initialReplayIdMap;
```
Comment (davsclaus) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java]: As its just a option for setting simple text=number, then we should make it a multi valued type, so you can configure this directly from the uris as well. See for example MailEndpoint for an example. You can then try to do this and do another PR with this improvement.

**Comment by davsclaus** [comment]
Thanks for the PR it has been merged. Do you mind closing this?

File: components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java (hunk 3 above)
Comment (sirlatrom) [inline components/camel-salesforce/camel-salesforce-component/src/main/java/org/apache/camel/component/salesforce/SalesforceEndpointConfig.java]: Is this all that is needed to set the value straight from the URI? https://github.com/apache/camel/blob/master/tooling/spi-annotations/src/main/java/org/apache/camel/spi/UriParam.java#L95

**Comment by dhirajsb** [comment]
Correct me if I am wrong, but the code for CometDReplayExtension was copied from Salesforce, right? What's their license for that code, and was copying it into the component unambiguously legal?

**Comment by sirlatrom** [comment]
@dhirajsb You're quite correct. While i haven't consulted with legal, the Salesforce consultants (from Salesforce inc.) embedded in our team have themselves recommended us to go ahead and use this class. I'll take up the issue with them ASAP.

**Comment by dhirajsb** [comment]
Thanks @sirlatrom. BTW, you're making great contributions to the Salesforce component, keep it going. :)

**Comment by sirlatrom** [comment]
@dhirajsb Salesforce have volunteered to license the repo I copied the class from under [BSD](https://github.com/developerforce/StreamingReplayClientExtensions/blob/master/LICENSE), but do you know if that is viable in the Apache Camel project?

**Comment by davsclaus** [comment]
You can see the list of approved / rejected licenses here
http://www.apache.org/legal/resolved.html#category-a

**Comment by sirlatrom** [comment]
Yay, I see http://opensource.org/licenses/BSD-3-Clause there :-)

**Comment by vijaychd** [comment]
when this will be released?
